core/controllers/email_dashboard.py

```diff
@@ -41,79 +43,59 @@ def test_user(self, **kwargs):
     return test_user
 
 
-class QueryPage(base.BaseHandler):
+class EmailDashboardPage(base.BaseHandler):
     """Page to submit query and show past queries."""
 
     @require_valid_sender
     def get(self):
         """Handles GET requests."""
-        self.render_template('pages/query/query.html')
+        self.render_template('pages/email_dashboard/email_dashboard.html')
 
 
-class QueryDataHandler(base.BaseHandler):
+class EmailDashboardDataHandler(base.BaseHandler):
     """Query data handler."""
 
+    QUERIES_PER_PAGE = 10
+
     @require_valid_sender
     def get(self):
-        offset = (
-            int(self.request.get('offset')) if self.request.get('offset') != ''
-            else 0)
-        base_offset = (
-            int(self.request.get('base_offset'))
-            if self.request.get('base_offset') != '' else 0)
-
-        query_models = (
+        cursor = Cursor(urlsafe=self.request.get('cursor'))
+
+        query_models, next_cursor, more = (
             user_models.UserQueryModel.query().
             order(-user_models.UserQueryModel.created_on).
-            fetch(base_offset, offset=offset))
-        queries_list = []
-
-        for model in query_models:
-            query = {}
-            query['id'] = model.id
-            query['submitter_id'] = user_services.get_username(
-                model.submitter_id)
-            query['created_on'] = model.created_on.strftime("%d-%m-%y %H:%M:%S")
-            query['status'] = model.query_status
-            query['qualified_users'] = len(model.user_ids)
-            queries_list.append(query)
-
-        self.render_json({
+            fetch_page(self.QUERIES_PER_PAGE, start_cursor=cursor))
+
+        submitters_settings = user_services.get_users_settings(
+            [model.submitter_id for model in query_models])
+
+        submitter_details = {
+            submitter.user_id: submitter.username
+            for submitter in submitters_settings
+        }
+
+        queries_list = [{
+            'id': model.id,
+            'submitter_id': submitter_details[model.submitter_id],
+            'created_on': model.created_on.strftime('%d-%m-%y %H:%M:%S'),
+            'status': model.query_status,
+            'num_qualified_users': len(model.user_ids)
+        } for model in query_models]
+
+        data = {
             'recent_queries': queries_list
-        })
+        }
+        if more and next_cursor:
+            data['cursor'] = next_cursor.urlsafe()
+
+        self.render_json(data)
 
     @require_valid_sender
     def post(self):
         """Post handler for query."""
         data = self.payload['data']
```

Comment:
'data' should be validated, and there should be tests to ensure that a 400 error is raised if invalid data is passed through.